src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -88,6 +105,108 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
+    ImmutableSet.Builder<ConfiguredTarget> incompatibleButRequestedTargets = ImmutableSet.builder();
+
+    for (ConfiguredTarget target : topLevelTargets) {
+      IncompatiblePlatformProvider incompatibleProvider =
+          target.getProvider(IncompatiblePlatformProvider.class);
+
+      // Move on to the next target if this one is compatible.
+      if (incompatibleProvider == null) {
+        continue;
+      }
+
+      // We need the label in unambiguous form here. I.e. with the "@" prefix for targets in the
+      // main repository. Otherwise we risk comparing two unrelated targets.
+      String labelString = target.getLabel().getUnambiguousCanonicalForm();
+      if (explicitTargetPatterns.contains(labelString)) {
+        if (!keepGoing) {
+          // Use the slightly simpler form for printing error messages. I.e. no "@" prefix for
+          // targets in the main repository.
+          String targetIncompatibleMessage =
+              ("Target "
+                  + target.getLabel().toString()
+                  + " is incompatible and cannot be built, but was explicitly requested.");
+          targetIncompatibleMessage += reportOnIncompatibility(target);
+          throw new ViewCreationFailedException(
+              targetIncompatibleMessage,
+              FailureDetail.newBuilder()
+                  .setMessage(targetIncompatibleMessage)
+                  .setAnalysis(Analysis.newBuilder().setCode(Code.INCOMPATIBLE_TARGET_REQUESTED))
+                  .build());
+        }
+        this.eventHandler.handle(
+            Event.warn(
+                "Incompatible target "
+                    + labelString
+                    + " was explicitly requested: it will not be built"));
+        incompatibleButRequestedTargets.add(target);
+      } else {
+        // If this is not an explicitly requested target we can safely skip it.
+        incompatibleTargets.add(target);
+      }
+    }
+
+    return new PlatformRestrictionsResult(
+        ImmutableSet.copyOf(incompatibleTargets.build()),
+        ImmutableSet.copyOf(incompatibleButRequestedTargets.build()));
+  }
+
+  /**
+   * Assembles the explanation for a platform incompatibility.
+   *
+   * <p>This is useful when trying to explain to the user why an explicitly requested target on the
+   * command line is considered incompatible. The goal is to print out the dependency chain and the
+   * constraint that wasn't satisfied so that the user can immediately figure out what happened.
+   *
+   * @param target the incompatible target that was explicitly requested on the command line.
+   * @return the verbose error message to show to the user.
+   */
+  private static String reportOnIncompatibility(ConfiguredTarget target) {
+    Preconditions.checkNotNull(target);
+
+    String message = "\nDependency chain:";
+    IncompatiblePlatformProvider provider = null;
+
+    while (target != null) {
+      message += "\n    " + target.getLabel().toString();
```

Comment:
Nit: The toString() call here and on line 205 isn't needed.

Reply:
Done